### PR TITLE
wallet: simplify pagination api

### PR DESCRIPTION
**Pull request by yyforyongyu**
## Summary
- remove builder-style pagination request methods and require explicit, non-zero page limits
- drop shared default and max-limit policy from `page`, leaving query sizing and validation to wallet/address store code
- simplify manual pagination by using pointer-style `After`/`Next` tokens and update tests/docs to match

## Why
The current pagination API still exposes more policy and plumbing than necessary for callers. This change keeps the shared `page` package focused on the pagination contract itself while letting stores own query sizing and validation. It also makes page-by-page call sites more direct by forwarding `Next` into `After` without extra helper or presence-flag state.

## Testing
- `GOWORK=off go test ./wallet/internal/db/page ./wallet/internal/db`
- `GOWORK=off go test -tags itest ./wallet/internal/db/itest -run 'TestListWallets|TestListWalletsZeroLimit|TestIterWallets|TestListAddresses|TestListAddressesZeroLimit|TestListAddressesOrdering|TestIterAddresses'`
- `GOWORK=off go test ./wallet/internal/db/...`

### CODE REVIEW

**Comment by yyforyongyu** [comment]
LGTM 🚀 (`gemini-3.1-pro-preview`)

**Comment by yyforyongyu** [comment]
LGTM 📄 (qwen3.5-397b-a17b)

No review threads yet. The PR cleanly simplifies the pagination API by:
- Removing builder-style methods (`WithLimit`, `WithAfter`) in favor of struct literals
- Making `After` a direct pointer field instead of optional cursor accessors
- Requiring explicit non-zero page limits (no implicit defaults)
- Dropping `QueryLimit()` and `normalizedLimit()` indirection
- Simplifying `BuildResult` to accept limit directly

All call sites updated correctly. Zero-limit validation added at store level with `ErrInvalidPageLimit`. Tests cover pagination edge cases including zero-limit rejection.

**Review by yyforyongyu — COMMENTED** [review]
LGTM 📄

Pagination semantics preserved — all 4 store implementations consistently check `Limit == 0`, compute `Limit + 1` for SQL, and pass `Limit` to `BuildResult`. Cursor pointer semantics identical to old `WithAfter`/`After()` accessors. `BuildResult` limit=0 panic path unreachable due to early guards. Removal of `DefaultLimit`/`MaxLimit` normalization is intentional (neither used outside `page` package). Tests updated and new zero-limit rejection tests added.

**(`claude-opus-4-6`)**

**Review by yyforyongyu — COMMENTED** [review]
LGTM 🌷 (`gpt-5.4`)

**Review by yyforyongyu — COMMENTED** [review]
LGTM ✅ (`gpt-5.3-codex`)

**Review by yyforyongyu — COMMENTED** [review]
Thanks for the pagination API cleanup! The simplification looks good - removing the builder-style methods and requiring explicit limits makes the API cleaner.

However, I have a concern about the missing upper bound validation mentioned in the open review threads. All four unresolved comments raise the same issue: while `Limit == 0` is now rejected, there's no upper bound check. A very large `Limit` could cause:
- Large DB queries (`LIMIT (Limit+1)`)
- High memory usage from materializing all rows before trimming

Should we add a store-specific maximum limit and return `ErrInvalidPageLimit` for limits exceeding it? This would prevent accidental or malicious large queries while keeping the policy in store code as intended by this PR. (`glm5`)

File: wallet/internal/db/wallet_sqlite.go [hunk 1]
```diff
@@ -145,6 +145,9 @@ func (s *SqliteStore) GetWallet(ctx context.Context,
 // ListWallets returns a page of wallets matching the given query.
 func (s *SqliteStore) ListWallets(ctx context.Context,
 	query ListWalletsQuery) (page.Result[WalletInfo, uint32], error) {
+	if query.Page.Limit == 0 {
+		return page.Result[WalletInfo, uint32]{}, ErrInvalidPageLimit
+	}
```
[the remark this answers is not in the document]
Comment (yyforyongyu) [reply wallet/internal/db/wallet_sqlite.go]: Intentional here. This change removes shared max-limit policy from the pagination layer and keeps only request-shape validation (`Limit > 0`) at the store boundary. If we decide specific call paths need caps, I would rather apply those at the caller-facing boundary (RPC/API/request source) instead of reintroducing a single global maximum in store pagination code.

File: wallet/internal/db/wallet_pg.go [hunk 2]
```diff
@@ -145,6 +145,9 @@ func (s *PostgresStore) GetWallet(ctx context.Context,
 // ListWallets returns a page of wallets matching the given query.
 func (s *PostgresStore) ListWallets(ctx context.Context,
 	query ListWalletsQuery) (page.Result[WalletInfo, uint32], error) {
+	if query.Page.Limit == 0 {
+		return page.Result[WalletInfo, uint32]{}, ErrInvalidPageLimit
+	}
```
[the remark this answers is not in the document]
Comment (yyforyongyu) [reply wallet/internal/db/wallet_pg.go]: Intentional here. This change removes shared max-limit policy from the pagination layer and keeps only request-shape validation (`Limit > 0`) at the store boundary. If we decide specific call paths need caps, I would rather apply those at the caller-facing boundary (RPC/API/request source) instead of reintroducing a single global maximum in store pagination code.

File: wallet/internal/db/addresses_sqlite.go [hunk 3]
```diff
@@ -36,14 +36,17 @@ func (s *SqliteStore) GetAddress(ctx context.Context,
 // ListAddresses returns a page of addresses matching the given query.
 func (s *SqliteStore) ListAddresses(ctx context.Context,
 	query ListAddressesQuery) (page.Result[AddressInfo, uint32], error) {
+	if query.Page.Limit == 0 {
+		return page.Result[AddressInfo, uint32]{}, ErrInvalidPageLimit
+	}
```
[the remark this answers is not in the document]
Comment (yyforyongyu) [reply wallet/internal/db/addresses_sqlite.go]: Intentional here. This change removes shared max-limit policy from the pagination layer and keeps only request-shape validation (`Limit > 0`) at the store boundary. If we decide specific call paths need caps, I would rather apply those at the caller-facing boundary (RPC/API/request source) instead of reintroducing a single global maximum in store pagination code.

File: wallet/internal/db/addresses_pg.go [hunk 4]
```diff
@@ -36,14 +36,17 @@ func (s *PostgresStore) GetAddress(ctx context.Context,
 // ListAddresses returns a page of addresses matching the given query.
 func (s *PostgresStore) ListAddresses(ctx context.Context,
 	query ListAddressesQuery) (page.Result[AddressInfo, uint32], error) {
+	if query.Page.Limit == 0 {
+		return page.Result[AddressInfo, uint32]{}, ErrInvalidPageLimit
+	}
```
[the remark this answers is not in the document]
Comment (yyforyongyu) [reply wallet/internal/db/addresses_pg.go]: Intentional here. This change removes shared max-limit policy from the pagination layer and keeps only request-shape validation (`Limit > 0`) at the store boundary. If we decide specific call paths need caps, I would rather apply those at the caller-facing boundary (RPC/API/request source) instead of reintroducing a single global maximum in store pagination code.

**Review by yyforyongyu — COMMENTED** [review]
LGTM 🎲

Thanks for the clarification. Makes sense - keeping limit policy at the caller-facing boundary (RPC/API) rather than in store pagination avoids reintroducing a global maximum. The API now cleanly separates concerns:
- Pagination layer: request shape validation (Limit > 0)
- Caller boundaries: domain-specific caps as needed

Clean simplification. (`glm5`)

File: wallet/internal/db/itest/address_store_test.go [hunk 5]
```diff
@@ -739,6 +738,7 @@ func TestListAddresses(t *testing.T) {
 					WalletID:    walletID,
 					Scope:       db.KeyScopeBIP0044,
 					AccountName: "test-account",
+					Page:        page.Request[uint32]{Limit: 10},
 				}
```
Comment (GustavoStingelin) [inline wallet/internal/db/itest/address_store_test.go]: So, we do not have a default value anymore? Should every caller specify the page size that he wants?

**Review by GustavoStingelin — APPROVED** [review]
I’m not convinced this change produces a better API.

We’re moving from a richer page package, which centrally enforces invariants and validation, to a more anemic struct that mainly carries values. In practice, that shifts responsibility to callers to validate inputs themselves, including cases like zero or out-of-range limits.

The tradeoff is that:
- validation logic is more likely to be repeated across callers
- behavior can become inconsistent if different callers apply different rules
- constraints such as database-specific maximum limits are harder to enforce centrally

Previously, those guarantees were encoded in the API itself. Now, correct behavior depends more on each caller using the type correctly.

I’m approving to keep things moving, but I think this direction is debatable and worth getting more feedback from other developers on the team before we treat it as a clear improvement.

**Comment by yyforyongyu** [comment]
> Previously, those guarantees were encoded in the API itself. Now, correct behavior depends more on each caller using the type correctly.

Agree that the earlier plain-data version pushed too much invariant enforcement out to each caller/store. That part of your concern was valid, and this follow-up is meant to address it by moving page-limit validation back into `page.NewRequest(...)` while still keeping the common pagination flow direct.

However, I still think max-limit policy is a different concern from request-shape validation, and I’m less convinced that it belongs in `page` itself. For instance, a user-facing RPC may want to cap page size at a few hundred items, while an internal maintenance or batch path may reasonably want a much larger bound. **A single global max in the pagination package would force one policy across all of those call sites.**

So the split I’m aiming for is: let page own the pagination contract and its invariants, while caller-facing boundaries decide whether a particular use case should impose an upper bound.

**Comment by GustavoStingelin** [comment]
> For instance, a user-facing RPC may want to cap page size at a few hundred items…

What do you think about having multiple caps instead of a single global one?

One closer to the database layer, focused on protecting query performance and resource usage. Another at the RPC layer, focused on fair usage and preventing overly heavy requests from callers.

I agree that a single global limit can be too restrictive for some paths. Layered limits seem more flexible, for example:

* a higher cap at the DB level, tuned for efficient execution
* a stricter cap at the RPC level, aligned with API expectations

Alternatively, caps could be defined per query, depending on data shape and performance characteristics, rather than enforcing a one size fits all limit.

File: wallet/internal/db/page/request.go [hunk 6]
```diff
@@ -1,95 +1,50 @@
...
-// A false ok return value means the first page is being requested.
-func (r Request[Cursor]) After() (Cursor, bool) {
-	if r.after == nil {
-		var zero Cursor
-
-		return zero, false
+// NewRequest returns a Request with the given positive page limit.
+func NewRequest[Cursor any](limit uint32) (Request[Cursor], error) {
+	if limit == 0 {
+		return Request[Cursor]{}, ErrInvalidLimit
```
Comment (GustavoStingelin) [inline wallet/internal/db/page/request.go:22]: I really like this approach, but I moved away from it because it still does not prevent someone from passing a zero-value `page.Request{}` as a parameter. So we would still end up needing validation in the callers, since we cannot enforce that the constructor is the only way to build that value.

But now, I have realized there may be a way to preserve the same idea. If we move `Request` from the `page` package into the `db` package, we could also make it unexported, for example `db.pageRequest`, and expose only `db.NewPageRequest`. That way, we would actually get the encapsulation benefits, because callers would no longer be able to instantiate the type directly and bypass normalization.

What do you think about this option?

Comment (yyforyongyu) [reply wallet/internal/db/page/request.go:22]: We could try it in code and see if we like it or not - my hunch is that, it will make the `page` too small, and it seems like the issue still exists but in a different place, that zero-value Query structs still embed zero-value `pageRequest`?
